### PR TITLE
EDU-1259 Add document revision comparison page

### PR DESCRIPTION
**Pull request by ahelmberger**
Closes https://educandu.atlassian.net/browse/EDU-1259

### CODE REVIEW

File: src/plugins/table-of-contents/table-of-contents-display.js [hunk 1]
```diff
@@ -17,15 +23,18 @@ export default function TableOfContentsDisplay({ content }) {
...
 
   return (
     <div className="TableOfContentsDisplay">
       <div className="TableOfContentsDisplay-content">
         {!!text && (
           <Markdown className="TableOfContentsDisplay-text">{text}</Markdown>
         )}
-        {!!nodes.length && (
+        {!!context.isPreview && (
+          <div className="TableOfContentsDisplay-previewWarning">{t('previewWarning')}</div>
```
Comment (adrianaluchian) [inline src/plugins/table-of-contents/table-of-contents-display.js:35]: Good idea!

File: src/components/pages/comparison.js [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+  const [comparison, setComparison] = useState(null);
+  const [oldRevisionId, setOldRevisionId] = useState(request.query.oldId);
+  const [newRevisionId, setNewRevisionId] = useState(request.query.newId);
+  const [displayModesByMappedRevisionKey, setDisplayModesByMappedRevisionKey] = useState({});
+
+  const mappedRevisions = useMemo(() => [...revisions.sort(by(x => x.order, 'desc'))].map((revision, index) => {
+    const isLatestVersion = index === 0;
+    const version = revisions.length - index;
+    const latestVersionText = isLatestVersion ? ` (${t('common:latest')})` : '';
+    const versionText = `${t('common:version')} ${version}${latestVersionText}`;
```
Comment (adrianaluchian) [inline src/components/pages/comparison.js:62]: Since we now have basically the exact same logic here and on `document-version-history`, but also a variation in the `getDocumentRevisionHistoryVersionInfo` helper used on the `document` page (which is still sorting incorrectly by `createdOn` and not by `order` 😬 ), I'm thinking we could transform the helper into a `getDocumentRevisionsWithVersionInfo` (or something of sorts) that maps `isLatestVersion` and `versionText` to the `documentRevision`. And use that in all 3 places. 
What do you think?

File: src/utils/document-revision-comparison-utils.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+  return result;
+}
+
+function getMetadataTextRepresentation(documentRevision) {
+  if (!documentRevision) {
+    return '';
+  }
+
+  const shallowClone = cloneDeep({ ...documentRevision });
```
Comment (adrianaluchian) [inline src/utils/document-revision-comparison-utils.js:93]: This is not a shallow clone actually, right? 🤔 And destructuring isn't needed, I think.

File: src/components/pages/comparison.js [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+    return renderPanel({ id: 'metadata', info: `${t('metadata')}`, actions: [], content: panelContent });
+  };
+
+  const renderSectionComparison = sectionComparison => {
+    if (sectionComparison.changeType === SECTION_CHANGE_TYPE.movedDown || sectionComparison.changeType === SECTION_CHANGE_TYPE.movedUp) {
+      return renderSectionMovementInfo(sectionComparison);
+    }
+
+    const displayMode = displayModesByMappedRevisionKey[sectionComparison.key];
```
Comment (adrianaluchian) [inline src/components/pages/comparison.js:221]: I'm not sure if it's this `key` or another, that causes a react warning of "two children with the same key" when previewing a section that contains changes. Does not occur for when the section was added or deleted.

**Comment by adrianaluchian** [comment]
This comparison page is so much better than I thought it would end up 😍 You managed to showcase this feature's value with the way you implemented it, congrats! 👏

File: src/utils/document-revision-comparison-utils.js (hunk 3 above)
Comment (ahelmberger) [reply src/utils/document-revision-comparison-utils.js:93]: HA! Seems like I first did the one, then the other, and ended up having both with the name of the other. If you're confused now, don't worry, it was intentional.

Switched to only having `cloneDeep` and the variable named `clonedRevision` (although shallow clone would have sufficed, it's dangerous if code is added later). I also realized when fixing this, that we should revert the user object expansion done by the `ClientDataMapper` for the `publicContext.accreditedEditors` array like we do for `createdBy`.

tldr; Fixed! Thanks for spotting it!

File: src/utils/document-revision-comparison-utils.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    const newSectionKey = newSectionKeys[indexInNewSections] ?? null;
+
+    if (oldSectionKey === newSectionKey) {
+      // section stayed in the same place
+      result.push([oldSectionKey, SECTION_CHANGE_TYPE.unchanged]);
+      processedSectionKeys.add(oldSectionKey);
+      indexInOldSections += 1;
+      indexInNewSections += 1;
+    } else if (newSectionKey !== null && !oldSectionKeySet.has(newSectionKey)) {
+      // section was added in `sectionKeys2`
```
[the remark this answers is not in the document]
Comment (ahelmberger) [reply src/utils/document-revision-comparison-utils.js:61]: That's indeed not very great. As I'm not a fan of `continue`s I extracted all variables on top ...

~~~
hasSectionBeenInTheSamePlace
hasSectionBeenAdded
hasSectionBeenDeleted
hasSectionBeenMovedToHere
hasSectionBeenMovedFromHere
~~~

... and used them in the `if`-`else if` as conditions. Of course I removed the comments :grin:

File: src/utils/document-revision-comparison-utils.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+      result.push([newSectionKey, SECTION_CHANGE_TYPE.movedHere]);
+      processedSectionKeys.add(newSectionKey);
+      indexInNewSections += 1;
+    } else if (oldSectionKey !== null && movedSectionKeySet.has(oldSectionKey)) {
+      // section was moved from here to a different place in `sectionKeys2`
+      result.push([oldSectionKey, processedSectionKeys.has(oldSectionKey) ? SECTION_CHANGE_TYPE.movedUp : SECTION_CHANGE_TYPE.movedDown]);
+      processedSectionKeys.add(oldSectionKey);
+      indexInOldSections += 1;
+    } else {
+      throw new Error('Unexpected case while creating section diff');
```
[the remark this answers is not in the document]
Comment (ahelmberger) [reply src/utils/document-revision-comparison-utils.js:82]: It would introduce yet another case that we have to render somehow and we also have already some decisions to make on the UI. This is the kind of `throw` that should never ever be executed unless the algorithm is wrong. I consider it more for developers to act like a compiler error or a guard in case of later changes. By introducing an `unknown` state we would promote the error to be part of the business logic. A section can either be added, deleted, moved, or stay put where it was (when it comes to placement inside a document). I cannot think of another unknown change type. So I would rather not open this door and have to support it throughout the UI (also: we couldn't even test it as it should never happen). So actually, if anything, I'd rather remove the `else` block completely. Whatcha think?

File: src/components/pages/comparison.js (hunk 4 above)
Comment (ahelmberger) [reply src/components/pages/comparison.js:221]: Good find, thanks. Fixed!

File: src/components/pages/comparison.js (hunk 2 above)
Comment (ahelmberger) [reply src/components/pages/comparison.js:62]: Sounds like a good idea! I'll extract a utility function.

Comment (ahelmberger) [reply src/components/pages/comparison.js]: Done!

Comment (adrianaluchian) [reply src/components/pages/comparison.js]: Perfect! 👏

File: src/utils/document-revision-comparison-utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrianaluchian) [reply src/utils/document-revision-comparison-utils.js:82]: I see your point 🤔 If it were the type of project where error logging were collected and reported to the dev team, this is completely how this would be done, while at the same time having the UI gracefully handle or swallow the scenario.
In our current setup, lots of such throws intended for the dev team are actually seen by the user (like "Was looking for x users but found y") cause we don't have this separation set up. 
So maybe we could combine both worlds a bit and instead of throwing we would console log? This way the user is oblivious to the case, but if they would notice something missing in the comparison and report it, we could ask for a screenshot of the browser console (or see it ourselves when we open the comparison).

File: src/utils/document-revision-comparison-utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrianaluchian) [reply src/utils/document-revision-comparison-utils.js]: Wonderful, thanks! 👍

File: src/utils/document-revision-comparison-utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahelmberger) [reply src/utils/document-revision-comparison-utils.js]: Good idea. Actually we are neglecting the possibility to log stuff (probably due to the reason you mentioned), so I switched from throwing to logging :+1:

**Comment by ahelmberger** [comment]
> This comparison page is so much better than I thought it would end up heart_eyes You managed to showcase this feature's value with the way you implemented it, congrats! clap

Thanks :bowing_man:

File: src/utils/document-revision-comparison-utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrianaluchian) [reply src/utils/document-revision-comparison-utils.js]: Nice 👍

**Review by adrianaluchian — APPROVED** [review]
